Concrete role types (PR 6885)

izmalk opened the pull request:
## What is the goal of this PR?

Document changes from the PR https://github.com/vaticle/typedb/pull/6885 

## What are the changes implemented in this PR?

Improve roles description, add Abstract role sub-section.
Fix some TypeQL examples indentation.

izmalk committed:
TypeQL code indentation

izmalk committed:
Add abstract roles note

izmalk committed:
Minor edits

izmalk commented on typeql-src/modules/ROOT/pages/schema/define-types.adoc:
```diff
@@ -219,7 +219,9 @@ The following `define` query creates a new entity type, `object`, by subtyping t
 [,typeql]
 ----
-define object sub entity;
+define
```
Just fixing indentation and formatting for TypeQL on this page

izmalk committed:
TypeQL indentation and minor edits

izmalk commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -72,16 +72,18 @@ A string or a file with multiple TypeQL queries is a script.
 $p isa person,
-  has full-name "Masako Holley",
-  has email "masako.holley@vaticle.com";
+    has full-name "Masako Holley",
```
Sorry, there are some indentation fixes.

izmalk committed:
Merge branch 'development' into pr-6885-concrete-role-types

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
 ====
-Roles can be used in queries as a part of a relation even though roles do not have a direct type definition.
+
+.See the example of a concrete role in abstract type
```
This is a weak reference, since it's hard to maintain its integrity? You can just leave it at the above note IMO

izmalk replied:
I'm not sure what you mean by "weak reference". 
This is the title of a collapsible block. It has to be specific enough for users to decide whether they want to expand the content or not. Given the special way we define and use roles, the above note is insufficient to explain what is going on, IMHO. The shortest way to go about the explanation is to go with a short and highlighted Note and an expandable explanation below.

flyingsilverfin replied:
If it's the title can we just say "The example below illustrates concrete roles in abstract relation types"?

izmalk replied:
It's a collapsible title. By default, it is collapsed. So, we have to use an invitation to lure users to click on it. 
I'm trying to use it consistently, starting with "See". For example, "See the example..." or "See examples".
We can discuss how to address these collapsible examples as a part of a major style guide discussion if you want.
Here is how it looks on the website: https://typedb.com/docs/typeql/2.x/fundamentals#_relation_types
<img width="454" alt="Screenshot 2023-09-06 at 17 09 19" src="https://github.com/vaticle/typedb-docs/assets/48120135/459791c4-0ed4-47d4-888c-0ff5ef0e2b36">
<img width="687" alt="Screenshot 2023-09-06 at 17 09 51" src="https://github.com/vaticle/typedb-docs/assets/48120135/ca5e93ae-323c-4a53-ba0b-9612143c363f">

flyingsilverfin replied:
Ok that's pretty nice!

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
+.See the example of a concrete role in abstract type
+[%collapsible]
+====
+For example, given the following schema:
```
Do you need "For example"?

izmalk replied:
```suggestion
Given the following schema:
```
Not so much. I usually mention examples inside the collapsible block just as an introductory sentence. We can drop it here.

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
+
+group-membership sub membership;
+----
+
```
Since this entire section is about the change before and after 2.22.0, I would lead with that throughout:
```
Before TypeDB 2.22.0, `group-membership` would inherit the abstract role `member` from the abstract relation `group-membership`. This required us to override the abstract role with a new concrete one before we can insert data against it.

Since TypeDB 2.22.0, `group-membership` would inherit the concrete role `member` and can use it directly without the need to override it with a new role:
```

izmalk replied:
I'd prefer to not repeat the common part of `group-membership` would inherit the abstract role `member`, as I did in the sentence below.

izmalk replied:
Fixed in https://github.com/vaticle/typedb-docs/pull/714/commits/303e68ed2aff65463035d2205890d009ff6a06d8

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
+The concrete relation type `group-membership` inherits the `member` role from the abstract relation
+type `membership`.
+
+Previously, that would make the role `membership` to be abstract. To insert an instance
```
What are "previously" and "now"? If we read this in 6 months and we're on version 2.28.0, these time-based terms aren't accurate - I would avoid them as they go stale.

izmalk replied:
The whole example is for change, announced to take effect in version 2.22. But you're right; we should mention the version here again. I'll fix it.

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
+The concrete relation type `group-membership` inherits the `member` role from the abstract relation
+type `membership`.
+
+Previously, that would make the role `membership` to be abstract. To insert an instance
```
Note on grammr: "That" is ambiguous - grammatically you're referring to the 'inheritance' of the role. However, the inheritance itself isn't what made the role abstract - it's that the relation was abstract and it was a previously version so it's not quite right.

izmalk replied:
> inherits the `member` role from the abstract relation

 makes the role abstract and is referred to as "that". 
It seems OK to me, but if you think it is not, I'll try to change the wording here and consult our copy editor on that case.

izmalk replied:
We discussed the issue with our copy editor and will add a note to our style guide. Thank you!

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
+
+[WARNING]
+====
+Old databases created in previous versions of TypeDB can retain the old behaviour with possible abstract roles.
```
Schemas in databases created in versions of TypeDB prior to 2.22.0 will continue to use abstract roles when originally defined in abstract relations.

izmalk replied:
It seems wrong to talk about schema in this context - as if a schema is using abstract roles. 
On top of that, if we export schema in TypeQL, it would be exactly the same in both versions (prior to 2.22.0 and in 2.22.0+). I know it's in the implementation that we can't see in TypeQL. It seems natural to talk about database behavior instead of schema here to avoid explaining unnecessary implementation details.

izmalk replied:
I have updated the Warning text. It should be better now.

izmalk replied:
Fixed in https://github.com/vaticle/typedb-docs/pull/714/commits/303e68ed2aff65463035d2205890d009ff6a06d8

flyingsilverfin commented on typeql-src/modules/ROOT/pages/fundamentals.adoc:
```diff
@@ -396,36 +400,100 @@ For example, given the schema:
+Old databases created in previous versions of TypeDB can retain the old behaviour with possible abstract roles.
+
+The best way to switch to the new behavior is to install the TypeDB v.2.22.0+, create a new database,
+define the same schema, and load the same data into it.
```
Do you want add an explicit 'contact us on on our discord if you must do an in-place upgrade'?

izmalk replied:
```suggestion
The best way to switch to the new behavior is to install the TypeDB v.2.22.0+, create a new database,
define the same schema, and load the same data into it.

Contact us on our https://typedb.com/discord[Discord] if you need to do an in-place upgrade.
```
Are you sure it's not easier to include the short guide on how to do that in an expandable block here? It's not that big.
Or we can provide a link to the release notes instead.

flyingsilverfin replied:
We can do a short expandable guide here, let's do that :)

izmalk replied:
Fixed in https://github.com/vaticle/typedb-docs/pull/714/commits/bcbae00e6d7a2bc475dce7dfc13ee77f3d1049b3
<img width="1489" alt="Screenshot 2023-09-06 at 17 54 04" src="https://github.com/vaticle/typedb-docs/assets/48120135/3cba2e0f-8ca2-4ab0-a762-3533c9547c8d">
<img width="895" alt="Screenshot 2023-09-06 at 17 53 47" src="https://github.com/vaticle/typedb-docs/assets/48120135/1d3b1d55-8498-486d-adb9-b9ba5322756a">

flyingsilverfin commented on typeql-src/modules/ROOT/pages/schema/define-types.adoc:
```diff
@@ -273,15 +281,20 @@ To add roles that entities of a specific entity type can play, use the `plays` k
 ====
-All types that are subtyping `entity` built-in type directly or through other subtypes are called entity types. Instances of these types are called entities. The same approach can be applied to attributes and relations.
+All types that are subtyping `entity` built-in type directly or through other subtypes are called entity types.
+Instances of these types are called entities. The same approach can be applied to attributes and relations.
```
Should "Approach" be "terminology"? -- eg "Attributes and relations use analogous terminology". Or something?

Is it "can" or "is/are/should be" (more assertive?)

izmalk replied:
```suggestion
Instances of entity types are called entities. A similar terminology is applied to attributes and relations.
```

izmalk committed:
Apply suggestions from code review

izmalk committed:
Update concrete role after review

Example rewrite, Warning update, Version syntax fixes.
Add a link to override a role.

izmalk committed:
Update typeql-src/modules/ROOT/pages/schema/define-types.adoc

izmalk committed:
Add expandable guide to manual abstract upgrade
